### ISSUE TITLE
GUNROCK - Hi-performance Graph processing engine

### ISSUE THREAD

**Issue by archenroot** [bytedeco/javacpp-presets#641]
Some time passed I got back to JavaCPP :-)

In my current scope there is now: https://gunrock.github.io/docs/

Is it somethink you might be interested at Skymind projects with deeplearning4j as well, in such case maybe you could help me with preset generation :-)

Let me know and all the best.

Ladislav

**Comment by saudet** [comment]
Welcome back :) It has a C API, that should be pretty easy to map. Give it a try with this new guide:
https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes

**Comment by archenroot** [comment]
@saudet - ok, I prepared project skeleton with cppbuil.sh and created general infomapper class, now I strugle little bit, but I will come with real questions and possibly multiple :-)

**Comment by archenroot** [comment]
@saudet  - I kicked off the module creation at: 
https://github.com/archenroot/javacpp-presets

I will go trought the output, there is just something related to some JSON mapping comming from gunrock which generated some mess. When you have minute, could you quickly check whats wrong. 

Additionally I don't much understand from documentation what the link property means. At some projects I look at it has value like "hdf5@1", wasn't able to find much explained about this.

Thx for kick me further to finish this piece of preset.

**Comment by archenroot** [comment]
@saudet - link is clear.

Regarding json, I will paste here what has been produced by javacpp, just jump on cigarette...

**Comment by archenroot** [comment]
@saudet  - Ok, so I successfully generated the JNI wrapper, but its non-compilable:
`gunrock.java:749: error: <identifier> expected
    @Namespace("json_spirit") @MemberGetter public static native @Const @ByRef @Name("Value_impl<Config>::null") json_spirit::Value_impl<Config> Value_impl<Config>::null();
                                                                                                                            ^
gunrock.java:749: error: ';' expected
    @Namespace("json_spirit") @MemberGetter public static native @Const @ByRef @Name("Value_impl<Config>::null") json_spirit::Value_impl<Config> Value_impl<Config>::null();
                                                                                                                                                           ^
gunrock.java:749: error: illegal start of type
    @Namespace("json_spirit") @MemberGetter public static native @Const @ByRef @Name("Value_impl<Config>::null") json_spirit::Value_impl<Config> Value_impl<Config>::null();
                                                                                                                                                                   ^
gunrock.java:749: error: '(' expected
    @Namespace("json_spirit") @MemberGetter public static native @Const @ByRef @Name("Value_impl<Config>::null") json_spirit::Value_impl<Config> Value_impl<Config>::null();
                                                                                                                                                                     ^
gunrock.java:749: error: illegal start of type
    @Namespace("json_spirit") @MemberGetter public static native @Const @ByRef @Name("Value_impl<Config>::null") json_spirit::Value_impl<Config> Value_impl<Config>::null();
                                                                                                                                                                         ^
gunrock.java:749: error: <identifier> expected
    @Namespace("json_spirit") @MemberGetter public static native @Const @ByRef @Name("Value_impl<Config>::null") json_spirit::Value_impl<Config> Value_impl<Config>::null();
                                                                                                                                                                          ^
6 errors
`
Current state is available here: https://github.com/archenroot/javacpp-presets

**Comment by saudet** [comment]
That doesn't look like the C API. Start with that, it'll be easier.

**Comment by archenroot** [comment]
Not it doesn't :-) - https://github.com/cierelabs/json_spirit (C++)

**Comment by saudet** [comment]
No, like this: https://github.com/gunrock/gunrock/tree/master/shared_lib_tests
Basically the "gunrock.h" include file only.

**Comment by archenroot** [comment]
I see, thx. Will start with that.

**Comment by archenroot** [comment]
Ok, mvn clean install sucess. I will add to the cppbuild.sh some condition to check if directory already exists, to prevent download and build, could be handy for debugging purposes.

**Comment by saudet** [comment]
The cppbuild.sh builds in the presets already don't redownload or rebuild, please use the same pattern!

Yes, if the C API is enough for your needs, let's go with that at first. I'm sure it will be useful to others, or if not maybe they'll help out and enhance it. :)

**Comment by archenroot** [comment]
Ok, my mistake, the problematic line should be:
`grGraphI.row_offsets(new IntPointer(row_offsets[0]));`

**Comment by archenroot** [comment]
@saudet - the example is working fine - it executes on GPU without any error, but from C example:
```
gunrock_sssp(grapho, graphi, config, data_t);

    ////////////////////////////////////////////////////////////////////////////
    int *labels = (int*)malloc(sizeof(int) * graphi->num_nodes);
    labels = (int*)grapho->node_value1;
    int node; for (node = 0; node < graphi->num_nodes; ++node)
        printf("Node_ID [%d] : Label [%d]\n", node, labels[node]);

```
I don't know how to get the actual array. grapho.node_value1 is type of Pointer, how to obtain an array which it points to? This is last piece of sample code to print out distance. And I am there thanks to you.

**Comment by saudet** [comment]
Cast it by creating a new IntPointer.

**Comment by archenroot** [comment]
I am playing around and being not able to find right way...

Case 1: 
```
Pointer labels = IntPointer.malloc(grGraphI.num_nodes());
        labels =  grGraphO.node_value1();
        IntPointer intl = new IntPointer(labels);
        int nodes = row_offsets.length - 1;
        for (int index =0;index < nodes;index++){
            log("distance:" + intl.get(index));
        }
``` produces:
```
input number of nodes:7
distance:0
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/util/array_utils.cuh, 209 @ gpu 0] keys cudaMalloc failed (CUDA error 2: out of memory)
distance:2147483647
distance:2147483647
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/sssp/sssp_app.cu, 200 @ gpu 0] SSSP Problem Enact Failed (CUDA error 2: out of memory)
distance:2147483647
distance:2147483647
distance:2147483647
distance:2147483647
```
Case 2:
`IntPointer labels = (IntPointer) IntPointer.malloc(grGraphI.num_nodes());`
produces:
```
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/enactor_loop.cuh, 781 @ gpu 0] FullQueue_Core failed. (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/sssp/sssp_app.cu, 200 @ gpu 0] SSSP Problem Enact Failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/util/array_utils.cuh, 550 @ gpu 0] distances cudaMemcpy D2H failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/sssp/sssp_app.cu, 209 @ gpu 0] SSSP Problem Data Extraction Failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/util/array_utils.cuh, 271 @ gpu 0] node_locks cudaFree failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/util/array_utils.cuh, 271 @ gpu 0] node_locks cudaFree failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/problem_base.cuh, 672 @ gpu 0] cudaEventDestroy failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/problem_base.cuh, 672 @ gpu 0] cudaEventDestroy failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/util/array_utils.cuh, 271 @ gpu 0] data_slices[] cudaFree failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/util/array_utils.cuh, 271 @ gpu 0] row_offsets cudaFree failed (CUDA error 77: an illegal memory access was encountered)
Exception in thread "main" java.lang.ClassCastException: org.bytedeco.javacpp.Pointer cannot be cast to org.bytedeco.javacpp.IntPointer
```
Case 3;
`IntPointer intl = new IntPointer(grGraphO.node_value1());`
produces:
```
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/enactor_loop.cuh, 781 @ gpu 0] FullQueue_Core failed. (CUDA error 77: an illegal memory access was encountered)
distance:1284984256
distance:32736
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/sssp/sssp_app.cu, 200 @ gpu 0] SSSP Problem Enact Failed (CUDA error 77: an illegal memory access was encountered)
distance:0
distance:0
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/util/array_utils.cuh, 550 @ gpu 0] distances cudaMemcpy D2H failed (CUDA error 77: an illegal memory access was encountered)
distance:0
distance:0
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/sssp/sssp_app.cu, 209 @ gpu 0] SSSP Problem Data Extraction Failed (CUDA error 77: an illegal memory access was encountered)
distance:0
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/util/array_utils.cuh, 271 @ gpu 0] node_locks cudaFree failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/util/array_utils.cuh, 271 @ gpu 0] node_locks cudaFree failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/problem_base.cuh, 672 @ gpu 0] cudaEventDestroy failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/problem_base.cuh, 672 @ gpu 0] cudaEventDestroy failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/util/array_utils.cuh, 271 @ gpu 0] data_slices[] cudaFree failed (CUDA error 77: an illegal memory access was encountered)
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/util/array_utils.cuh, 271 @ gpu 0] row_offsets cudaFree failed (CUDA error 77: an illegal memory access was encountered)
Round 1 of sssp.
```

**Comment by saudet** [comment]
I'm not sure why they allocate memory, assign it to `labels`, but then assign another pointer right away to `labels`. There's probably something wrong with that code...

**Comment by archenroot** [comment]
There is actually the python counter example I found which imports ctype lib:
```
gunrock.sssp(labels, preds, nodes, edges, row, col, val, 1, sources, 0)

### sample results
print ' sssp labels (distance):',
for idx in range(nodes): print labels[0][idx],
```

labels is type of Pointer javacpp in my case at moment. Maybe I might enforce it in the parser directly to IntPointer?

**Comment by archenroot** [comment]
I tried to rewrite now python example with sssp method:
```
public static void main(String[] args){
        //gunrock gunrock = new gunrock();
        gunrock.GRTypes grTypes = new gunrock.GRTypes();
        grTypes.VTXID_TYPE(VTXID_INT);
        grTypes.SIZET_TYPE(SIZET_INT);
        grTypes.VALUE_TYPE(VALUE_INT);
        int srcs[] = {0};

        gunrock.GRSetup grSetup = InitSetup(1, srcs);
        int num_nodes = 7; int num_edges = 15;
        //int row_offsets[] = readFile("row.txt");
        //int col_list[] = readFile("col.txt");
        //int val_list = readFile("val.txt");
        int row_offsets[]  = {0, 3, 6, 9, 11, 14, 15, 15};
        int col_indices[] = {1, 2, 3, 0, 2, 4, 3, 4, 5, 5, 6, 2, 5, 6, 6};
        int edge_values[] = {39, 6, 41, 51, 63, 17, 10, 44, 41, 13, 58, 43, 50, 59, 35};

        //Python sample based
        IntPointer row = new IntPointer(row_offsets.length);
        IntPointer col = new IntPointer(col_indices.length);
        IntPointer edg = new IntPointer(edge_values.length);
        int nodes = row_offsets.length -1;
        int edges = col_indices.length;

        IntPointer labels = new IntPointer(nodes);
        IntPointer preds = new IntPointer(nodes);
        IntPointer sources = new IntPointer(1);
        sssp(labels, preds, nodes, edges, row, col, edg, 1, sources, true);
        for (int index =0;index < nodes;index++){
            log("distance:" + labels.get(index));
        }
```
Its also throwing some errors:
```
distance:2147483647
distance:2147483647
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/sssp/sssp_problem.cuh, 594 @ gpu 0] SSSPProblem cudaMemcpy frontier_queues failed (CUDA error 11: invalid argument)
distance:2147483647
distance:2147483647
distance:2147483647
distance:2147483647
[/home/zangetsu/proj/private/gpu/javacpp-presets/gunrock/cppbuild/linux-x86_64/gunrock/gunrock/app/sssp/sssp_app.cu, 193 @ gpu 0] SSSP Problem Data Reset Failed (CUDA error 11: invalid argument)
distance:2147483647
Round 1 of sssp.
```

**Comment by saudet** [comment]
Allocating arrays with lengths `row_offsets.length`, etc won't initialize them with any values...

**Comment by saudet** [comment]
The same with the other example, those there should be something like:
```java
grGraphI.row_offsets(new IntPointer(row_offsets));
...
```

**Comment by archenroot** [comment]
Yeah I got it, hm, actually I do now python vs c code result merging as they differ on same set, could be some graph algo customization sssp vs gunrock_sssp, but that is off topic. 

The important thing is:
```
Current dir: /home/zangetsu/proj/private/gpu/javacpp-presets_archenroot
Current dir: /home/zangetsu/proj/private/gpu/javacpp-presets_archenroot
Current dir: /home/zangetsu/proj/private/gpu/javacpp-presets_archenroot
input number of nodes:7
distance:0
distance:3
distance:4
distance:5
distance:10
distance:13
distance:16
Round 1 of sssp.
0	 1	 0	 queue3 	 oversize :	 10 ->	 14
```
JavaCPP rocks :100:

**Comment by archenroot** [comment]
@saudet - Could you review: https://github.com/archenroot/javacpp-presets?organization=archenroot&organization=archenroot

Or should I do merge requests, etc. The library requires CUDA and Boost 1.58-1.6* installed. Is there any way for such kind of dependencies to enforce by JavaCPP (it might be hard as per distro package manager...)

**Comment by saudet** [comment]
We can install Boost in the cppbuild.sh file as with Caffe, and for CUDA just make sure install-travis.sh downloads and installs it.

**Comment by archenroot** [comment]
@saudet - will check and update it.
